sound/soc/sof/intel/mtl.c

```diff
@@ -612,6 +613,42 @@ static int mtl_dsp_disable_interrupts(struct snd_sof_dev *sdev)
 	return mtl_disable_interrupts(sdev);
 }
 
+static int mtl_dsp_pcm_delay(struct snd_sof_dev *sdev, struct snd_soc_component *component,
+			     struct snd_pcm_substream *substream)
+{
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct hdac_stream *hstream = substream->runtime->private_data;
+	struct snd_soc_component *scomp = sdev->component;
+	struct sof_ipc_stream_posn *posn;
+	struct snd_sof_pcm *spcm;
+	u32 host;
+	u32 llp_l, llp_u;
+
+	spcm = snd_sof_find_spcm_dai(scomp, rtd);
+	if (!spcm) {
+		dev_err(sdev->dev, "warn: can't find PCM with DAI ID %d\n", rtd->dai_link->id);
+		return -EINVAL;
+	}
+
+	posn = &spcm->stream[substream->stream].posn;
+	posn->wallclock = snd_sof_dsp_read64(sdev, HDA_DSP_BAR, MTL_HFDSPWC);
+	host = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, AZX_REG_VS_SDXDPIB_XBASE +
+				(AZX_REG_VS_SDXDPIB_XINTERVAL *
+				hstream->index));
+	if (host >= hstream->bufsize)
+		host = 0;
+
+	posn->host_posn = bytes_to_frames(substream->runtime, host);
```

Comment:
I don't think that makes sense.

The delay is calculated assuming the hw_ptr is determined previously, see e.g. in pcm_native.c

````
	status->appl_ptr = runtime->control->appl_ptr;
	status->hw_ptr = runtime->status->hw_ptr;
	status->avail = snd_pcm_avail(substream);
	status->delay = snd_pcm_running(substream) ?
		snd_pcm_calc_delay(substream) : 0;
````

If you read a NEW value for DPIB, then the math is not quite correct, is it? You have to use runtime->status->hw_ptr; IMHO

Reply:
````c
static inline snd_pcm_uframes_t
snd_pcm_calc_delay(struct snd_pcm_substream *substream)
{
	snd_pcm_uframes_t delay;

	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
		delay = snd_pcm_playback_hw_avail(substream->runtime);
	else
		delay = snd_pcm_capture_avail(substream->runtime);
	return delay + substream->runtime->delay;
}
`````
@plbossart  in my opinion,  we should calculate delay for  above "substream->runtime->delay". 

one reference of runtime->delay is like
````c
static unsigned int azx_get_pos_fifo(struct azx *chip, struct azx_dev *azx_dev)
{
     ...
	runtime->delay = AMD_FIFO_SIZE;
        ...
}
````

Reply:
@kv2019i please check my above comment.  I agree "the delay is calculated in relation to avail/hw_ptr/appl_ptr , which is also reported to application."  but we should calculate substream->runtime->delay only.

Reply:
@RanderWang the delay has to be computed as a difference between runtime->status->hw_ptr and link position. we cannot refresh the hw_ptr  in the delay calculation